Improve error handling for distributed autograd engine.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27940 Improve error handling for distributed autograd engine.**

1) If we receive an error for outstanding rpcs in distributed autograd, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.
3) Use std::weak_ptr for GraphTasks within NodeTasks. This would help in avoiding executing NodeTasks whose associated GraphTask has expired.
4) Added a `clean_shutdown` parameter to `dist_init` to support test cases which can't shutdown cleanly across all nodes (ex: simulating node failures).

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

Commit:
Improve error handling for distributed autograd engine.

1) If we receive an error for outstanding rpcs, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/engine.h:
```diff
@@ -94,13 +94,22 @@ struct GraphTask {
 
-  GraphTask(bool keep_graph, bool grad_mode, int reentrant_depth)
+  // Whether or not to stop execution for this GraphTask when an error is
+  // encountered.
```
What does it mean to "stop" execution for the GraphTask? You can't immediately shut down all the threads; do you also clear out all the ready queues in this case? (What if someone reinitiates a backwards?)

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -480,16 +484,21 @@ static variable_list call_function(NodeTask& task) {
+  {
+    // Lock mutex for writing to task.base_->captured_vars_ and looking up
+    // exec_info_.
+    std::lock_guard<std::mutex> lock(task.base_->mutex_);
```
Hmm, what's going on here? I see the updated comment, but I don't know why it used to be OK to access `exec_info_` without a lock and now it's not. (Or is this a bugfix?)

ezyang replied:
> I'm assuming the code had the invariant of read-only exec_info_ for some reason.

Not sure what you mean by this :)

ezyang replied:
Hmm, if we follow my suggestion at https://github.com/pytorch/pytorch/pull/27940#discussion_r334992862 you don't have to write to `exec_info_` anymore. So then we can undo this!

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_context.cpp:
```diff
@@ -1,10 +1,30 @@
+
+  torch::autograd::variable_list apply(
+      torch::autograd::variable_list&& grads) override {
+    throw error_;
```
This is the same thing as

```
// Identity in forward, Error in backward. Used to implement @once_differentiable
struct TORCH_API DelayedError : public Node {
  DelayedError(std::string msg, int num_inputs)
    : msg(std::move(msg)) {
      for (int i = 0; i < num_inputs; i++)
        add_input_metadata(Node::undefined_input());
    }

  variable_list apply(variable_list&& inputs) override;

  std::string msg;
};

```

in `torch/csrc/autograd/functions/basic_ops.h`

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_context.cpp:
```diff
@@ -94,10 +114,35 @@ void DistAutogradContext::setGraphTask(
+    std::string err(message.payload().begin(), message.payload().end());
+    auto exception = std::runtime_error(err);
+
+    // Enqueue 'ErrorFunc' on the local autograd engine.
```
This seems like a really circuitous way to do this. Why not just expose a method on `graphTask_` that triggers an error? No need to create a synthetic task...

pritamdamania87 replied:
I agree your proposal is a much simpler way to do this.

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_context.cpp:
```diff
@@ -94,10 +114,35 @@ void DistAutogradContext::setGraphTask(
+
+    // Increment out standing tasks for this function and set appropriate
+    // exec_info_ for this function to execute.
+    graphTask_->outstanding_tasks_++;
```
I am wondering if it is possible for `graphTask_->outstanding_tasks_ == 0` here. Whether or not this is possible depends on the relationship between outstanding rpcs and outstanding tasks. Can this invariant get clarified here?

pritamdamania87 replied:
I don't think we need the logic here if we just set an error on the graphTask.

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -418,8 +423,9 @@ def test_backward_autograd_engine_error(self):
-    @dist_init
-    @unittest.skip("Skipping this test temporarily since ProcessGroupAgent does not report errors on node failures")
+    @dist_init(sync_at_shutdown=False)
+    @unittest.skipIf(TEST_CONFIG.rpc_backend == RpcBackend.PROCESS_GROUP,
```
What errors do you expect `ProcessGroupAgent` to report? Will timeout on send be sufficient?

pritamdamania87 commented:
@ezyang I think your proposal of just setting an error on the graph task would be much better. Although, in general there is an issue with this PR when `exit_on_error` is true. The problem is that with `exit_on_error`, engine::Execute() returns to the user and this in turn destroys the GraphTask object. Although, there could still be NodeTask objects in the ready_queue which are holding a reference to the GraphTask (GraphTask*). When these tasks eventually execute, we would have a segfault. There are a few ways to fix this:

1. NodeTask holds a shared_ptr to the GraphTask keeping it alive as a result. This would work, but I'm not sure about the performance overhead here because the autograd engine might create a lot of NodeTasks.
2. We drain all the NodeTasks from the `ready_queue` before we return from engine::Execute. I'm not sure how to do this currently since the `ready_queue` is not indexed in anyway by a GraphTask and has a bunch of NodeTasks from different GraphTasks.
3. GraphTasks holds a list of NodeTasks for that GraphTask. When the GraphTask object is destroyed, it marks all the NodeTasks in its list with a flag which indicates these NodeTasks would not be executed.

ezyang commented:
Yes, this makes sense. We need to implement some sort of orderly shutdown. One thing I observe is that we should drain the ready queues no matter what. However, as you point out in (2), because there may be jobs from multiple graph tasks in the ready queue, so we can't conveniently drain them. This suggests to me that tombstoning the GraphTask is some way is the right thing to do (1) (and (3) seems annoying to me). This could be done either by making everything a `shared_ptr`, as you suggest, or you could have just the main thread have a `shared_ptr` and the other threads have `weak_ptr` (expired graph task means you don't have to run the node task anymore).

I am wondering now how the autograd engine handles this sort of situation today, even in the absence of distributed autograd. It would probably be worth seeing what happens here. Maybe @albanD knows

pritamdamania87 commented:
I think the shared_ptr + weak_ptr approach sounds best, I'll implement that in this PR.

> I am wondering now how the autograd engine handles this sort of situation today, even in the absence of distributed autograd. It would probably be worth seeing what happens here. Maybe @albanD knows

This works in the autograd engine today since it doesn't return until all outstanding tasks are done and hence there is a guarantee that no NodeTasks are in the queue for that GraphTask. We introduced `exit_on_error` in this PR and that breaks the assumption that no NodeTasks are in the queue for that GraphTask.

ezyang commented:
> This works in the autograd engine today since it doesn't return until all outstanding tasks are done and hence there is a guarantee that no NodeTasks are in the queue for that GraphTask. We introduced exit_on_error in this PR and that breaks the assumption that no NodeTasks are in the queue for that GraphTask.

Makes sense!

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs in distributed autograd, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.
3) Use std::weak_ptr for GraphTasks within NodeTasks. This would help in avoiding executing NodeTasks whose associated GraphTask has expired.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs in distributed autograd, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.
3) Use std::weak_ptr for GraphTasks within NodeTasks. This would help in avoiding executing NodeTasks whose associated GraphTask has expired.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

pritamdamania87 commented:
@ezyang I'm implemented the weak_ptr approach in this PR, could you take another look? Thanks!

ezyang commented on torch/csrc/autograd/engine.h:
```diff
@@ -152,7 +166,7 @@ struct TORCH_API Engine {
   // machinery and shouldn't be exposed to users in anyway.
   variable_list execute_with_graph_task(
-      GraphTask& graph_task,
+      std::shared_ptr<GraphTask> graph_task,
```
If this isn't actually going to take ownership, you should pass this by reference to avoid refcount bump

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_context.cpp:
```diff
@@ -94,10 +96,23 @@ void DistAutogradContext::setGraphTask(
 
 void DistAutogradContext::addOutstandingRpc(
     const std::shared_ptr<rpc::FutureMessage>& futureMessage) {
+  futureMessage->addCallback(std::bind(
```
If possible, we should prefer using lambdas instead of `std::bind`

ezyang replied:
https://stackoverflow.com/questions/17363003/why-use-stdbind-over-lambdas-in-c14

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_context.cpp:
```diff
@@ -94,10 +96,23 @@ void DistAutogradContext::setGraphTask(
+    // If we have an error, let the local autograd engine know about it.
+    std::string err(message.payload().begin(), message.payload().end());
+    graphTask_->set_exception(
+        std::make_exception_ptr(std::runtime_error(err)), nullptr);
```
It seems a bit weird that you copy into a `std::string`, and then pass a const reference to the string to the exception constructor which will do another copy

ezyang commented on torch/csrc/autograd/engine.h:
```diff
@@ -179,15 +193,23 @@ struct TORCH_API Engine {
+  void evaluate_function(
+      std::shared_ptr<GraphTask> graph_task,
+      Node* func,
+      InputBuffer& inputs);
```
Is this diff complete? I expected to see a change to the implementation of `evaluate_function` but I see no other occurrences of `evaluate_function` in this PR

ezyang replied:
Still not fixed. This might be a ghexport problem. I'll go look at the phab diff

pritamdamania87 replied:
Hmm, this is really weird. I double checked that the PR was exported correctly. But now magically it shows "no changes" for engine.cpp.

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs in distributed autograd, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.
3) Use std::weak_ptr for GraphTasks within NodeTasks. This would help in avoiding executing NodeTasks whose associated GraphTask has expired.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

ezyang commented:
I feel we have hit some sort of GitHub bug. I can see engine.cpp in the list of files changed but GitHub claims there are no changes. And when I look at the base commit and the head commit, I do see the change. Maybe you should just close this PR and export to a new one.

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs in distributed autograd, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.
3) Use std::weak_ptr for GraphTasks within NodeTasks. This would help in avoiding executing NodeTasks whose associated GraphTask has expired.

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -152,15 +152,21 @@ struct ReadyQueue {
-  return base_->reentrant_depth_;
+  std::shared_ptr<GraphTask> graph_task;
+  TORCH_INTERNAL_ASSERT(
+      graph_task = base_.lock(), "GraphTask is no longer valid!")
```
No, please do not do this. Although it is not the case now, we reserve the right to completely eliminate `TORCH_INTERNAL_ASSERT`. If we did that, you wouldn't get the graph task at all in that case.

ezyang replied:
Thinking about the error condition here, probably once we start processing a NodeTask, we should maintain liveness on the GraphTask. This implies to me that we shouldn't have this method on NodeTask at all; we should just ask for reentrant depth directly from our (live) GraphTask. Whether or not we should fix this in the PR, I need to read more to see.

pritamdamania87 replied:
One of the other concerns here is performance since `getReentrantDepth` is called as part of `CompareNodeTaskTime` which is the comparison method of the heap. This method would be called multiple times for each `push`. I'm not sure of a clean way to resolve this issue yet.

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -152,15 +152,21 @@ struct ReadyQueue {
     std::lock_guard<std::mutex> lock(mutex_);
+    std::shared_ptr<GraphTask> graph_task;
+    TORCH_INTERNAL_ASSERT(
+        graph_task = item.base_.lock(), "GraphTask is no longer valid!")
```
Hot take: `push()` should take `GraphTask*` as an argument. If I'm pushing a task onto the ready queue, the graph task ought to be live.

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -244,7 +250,7 @@ auto Engine::thread_init(int device) -> void {
   // better.
   set_device(device);
-  thread_main(nullptr);
+  thread_main(nullptr, false);
```
nit: `/* reentrant_thread */ false`

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -261,51 +267,72 @@ auto Engine::thread_init(int device) -> void {
+    if (!(local_graph_task = task.base_.lock())) {
+      // Reentrant thread's graph task should not expire since we hold a
+      // reference to it in this method.
+      TORCH_INTERNAL_ASSERT(!reentrant_thread);
```
Good

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -330,24 +357,44 @@ void Engine::reentrant_thread_init() {
     set_device(graph_task->owner_);
     total_depth = graph_task->reentrant_depth_;
-    thread_main(graph_task);
+    thread_main(graph_task, /* reentrant thread*/ true);
```
On second thought, I don't think `thread_main` should keep a graph task alive. Consider the following situation: you have a thread which reentrantly calls backwards. You spawn a reentrant thread which should return once the sub-backwards call is run. But then this sub-backwards has an error in one of its nodes, causing the GraphTask to go dead. We should stop executing all of the other nodes and raise an error through Python land. But you won't do this because the reentrant thread is keeping the graph task live.

pritamdamania87 replied:
If the reentrant thread hits an error it would set an appropriate error on the graph task and eventually when all work is done we would still report this error to the user. 

> We should stop executing all of the other nodes and raise an error through Python land

We do this only in the case of dist_engine since we set `exit_on_error` only in that case. In our current autograd engine, we would try to execute all other outstanding tasks and return an error to the user once all tasks have completed.

ezyang replied:
> If the reentrant thread hits an error it would set an appropriate error on the graph task and eventually when all work is done we would still report this error to the user.

This is true. But if we're not going to take care about not keeping graphs unnecessarily live, we might as well not use `weak_ptr` (and save ourselves some bugs).

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -661,48 +725,52 @@ void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
 
+bool graph_task_completed(const GraphTask& graph_task) {
+  return graph_task.outstanding_tasks_.load() == 0 ||
+      (graph_task.exit_on_error_ && graph_task.has_error_.load());
```
Now I am wondering... when should we ever NOT exit on error?

pritamdamania87 replied:
Well, I didn't want to change the behavior of the current autograd engine and hence I added this logic. I thought there was a good reason we continue executing other tasks even though we hit an error for one.

ezyang replied:
TBH, I think it's an oversight in the current engine. @albanD, @apaszke any ideas?

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -804,7 +872,7 @@ auto Engine::start_threads() -> void {
 }
 
-void Engine::add_thread_pool_task(GraphTask *graph_task) {
+void Engine::add_thread_pool_task(std::weak_ptr<GraphTask> graph_task) {
```
Surely you're not taking ownership of the weak pointer here?

pritamdamania87 replied:
Yes, but we do take ownership before we call `thread_main`. Didn't want to take ownership here since the task could be in the queue for a while.

ezyang commented:
This looks good. There are some minor blocking things. There is a more major question of whether or not the asserted weak pointer lock accesses are sound; in the comments above I've suggested api changes that could make them provably sound. I am also curious why testing if graph task is not nullptr is no longer sufficient to tell if you are a reentrant thread; maybe you just made this change for clarity purposes? Would be good to know.

pritamdamania87 commented:
> I am also curious why testing if graph task is not nullptr is no longer sufficient to tell if you are a reentrant thread; maybe you just made this change for clarity purposes?

Yes, I just felt it adding a `reentrant_thread` bool made it easier to read the code.

Commit:
Update on "Improve error handling for distributed autograd engine."


1) If we receive an error for outstanding rpcs in distributed autograd, we enqueue an appropriate error
on the local autograd engine.
2) Add an `exit_on_error` mode for the local autograd engine, where the
computation stops if we see an error.
3) Use std::weak_ptr for GraphTasks within NodeTasks. This would help in avoiding executing NodeTasks whose associated GraphTask has expired.
4) Added a `clean_shutdown` parameter to `dist_init` to support test cases which can't shutdown cleanly across all nodes (ex: simulating node failures).

Differential Revision: [D17916844](https://our.internmc.facebook.com/intern/diff/D17916844/)

[ghstack-poisoned]

albanD reviewed (commented):
@pritamdamania87 just reviving this PR as it was simpler to check how `TORCH_INTERNAL_ASSERT(!reentrant_thread);` could be triggered. Other internal assert might need to be removed and properly handled as well.

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -152,15 +152,19 @@ struct ReadyQueue {
 int NodeTask::getReentrantDepth() const {
-  return base_->reentrant_depth_;
+  std::shared_ptr<GraphTask> graph_task = base_.lock();
+  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!")
```
You are right and thats why we added this later on to fix this problem: https://github.com/pytorch/pytorch/blob/877ab3afe33eeaa797296d2794317b59e5ac90f4/torch/csrc/autograd/engine.cpp#L177.

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -244,7 +248,8 @@ auto Engine::thread_init(int device) -> void {
   // better.
   set_device(device);
-  thread_main(nullptr);
+  std::shared_ptr<GraphTask> graph_task = nullptr;
```
I think we can avoid doing this, I guess what happened was in a previous version of the PR the GraphTask was passed in as a non-const ref, so we couldn't just pass nullptr here. But with a const-ref, it should be fine passing in nullptr.

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -261,51 +266,72 @@ auto Engine::thread_init(int device) -> void {
+    if (!(local_graph_task = task.base_.lock())) {
+      // Reentrant thread's graph task should not expire since we hold a
+      // reference to it in this method.
+      TORCH_INTERNAL_ASSERT(!reentrant_thread);
```
This is correct, the assumption we had here was that the reentrant thread was only executing tasks from the GraphTask passed in, but this is not true when I look at the code more closely. I'm guessing just removing this check would suffice?

albanD replied:
I think we just want to ignore such tasks yes, that would correspond to "draining" that queue from the work associated with this graph_task.
